File: test/scripts/event_scripts.test.mjs
/**
 * End-to-end tests for the event/cohort issue→PR scripts.
 *
 *   npm test     (node --test)
 *
 * Each case runs the real script as a child process with an ISSUE_BODY that a
 * hostile submitter could have typed, a throwaway checkout as its working
 * directory and $GITHUB_OUTPUT pointed at a temp file — the scripts read only
 * env and the filesystem, so nothing here touches the GitHub API or the repo.
 *
 * The properties under test:
 *   * a `..` year or a traversing event id is refused and writes nothing;
 *   * a `### heading` typed inside a free-text answer does not win;
 *   * control characters in a name survive the YAML round trip;
 *   * no answer can forge an extra `$GITHUB_OUTPUT` line.
 */

import assert from 'node:assert/strict';
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import test from 'node:test';
import { spawnSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';

import * as yaml from 'js-yaml';

const SCRIPTS = path.join(path.dirname(fileURLToPath(import.meta.url)), '..', '..', 'scripts');

/** Minimal checkout: one cohort with one event page, plus an unrelated page a
 * traversal would land on. Returns the temp root. */
function fixtureTree() {
  const root = fs.mkdtempSync(path.join(os.tmpdir(), 'event-scripts-'));
  const eventDir = path.join(root, 'cohorts', '2026', 'events', 'kickoff');
  fs.mkdirSync(eventDir, { recursive: true });
  fs.writeFileSync(
    path.join(eventDir, 'index.md'),
    [
      '---',
      'layout: event',
      'title: "Cohort kickoff"',
      'cohort: "2026"',
      'event_id: kickoff',
      '---',
      '',
      'Body.',
      '',
    ].join('\n'),
    'utf8'
  );
  fs.mkdirSync(path.join(root, 'about'), { recursive: true });
  fs.writeFileSync(
    path.join(root, 'about', 'index.md'),
    '---\nlayout: page\ntitle: About\n---\n\nReal page.\n',
    'utf8'
  );
  fs.mkdirSync(path.join(root, '_data', 'cohorts'), { recursive: true });
  fs.writeFileSync(
    path.join(root, '_data', 'cohorts', '2026.yml'),
    'year: 2026\nevents:\n  - id: kickoff\n    name: "Cohort kickoff"\n    date: 2026-09-09\n',
    'utf8'
  );
  return root;
}

/**
 * Run one script against a fixture tree.
 * @returns {{status: number, stdout: string, stderr: string, raw: string, outputs: Map<string, string>}}
 */
function run(script, { body = '', title = '', root }) {
  const outputFile = path.join(root, `outputs-${Math.random().toString(36).slice(2)}.txt`);
  fs.writeFileSync(outputFile, '');
  const result = spawnSync(process.execPath, [path.join(SCRIPTS, script)], {
    cwd: root,
    encoding: 'utf8',
    env: { ...process.env, ISSUE_BODY: body, ISSUE_TITLE: title, GITHUB_OUTPUT: outputFile },
  });
  const raw = fs.readFileSync(outputFile, 'utf8');
  return {
    status: result.status,
    stdout: result.stdout,
    stderr: result.stderr,
    raw,
    outputs: parseOutputs(raw),
  };
}

/** Read a $GITHUB_OUTPUT file the way the Actions runner does. */
function parseOutputs(raw) {
  const outputs = new Map();
  const lines = raw.split('\n');
  for (let i = 0; i < lines.length; i += 1) {
    const heredoc = /^([A-Za-z0-9_-]+)<<(.+)$/.exec(lines[i]);
    if (heredoc) {
      const [, key, delimiter] = heredoc;
      const buffer = [];
      i += 1;
      while (i < lines.length && lines[i] !== delimiter) {
        buffer.push(lines[i]);
        i += 1;
      }
      outputs.set(key, buffer.join('\n'));
      continue;
    }
    const plain = /^([A-Za-z0-9_-]+)=(.*)$/.exec(lines[i]);
    if (plain) outputs.set(plain[1], plain[2]);
  }
  return outputs;
}

/** Front matter of a generated page, parsed. */
function frontMatterOf(file) {
  const text = fs.readFileSync(file, 'utf8');
  const match = /^---\n(.*?)\n---\n/s.exec(text);
  assert.ok(match, `no front matter in ${file}`);
  return yaml.load(match[1]);
}

const section = (heading, value) => `### ${heading}\n\n${value}\n\n`;

// --- new_event_from_issue.mjs ----------------------------------------------

test('new_event: a heading inside the details answer cannot replace the event id', () => {
  const root = fixtureTree();
  const body =
    section('Cohort Year', '2026') +
    section('Event Title', 'Real title') +
    section('Event ID', 'orientation') +
    section('Event Details (Markdown, optional)', 'Notes below.\n\n### Event ID\n\nhijacked');

  const result = run('new_event_from_issue.mjs', { body, root });

  assert.equal(result.status, 0, result.stderr);
  assert.equal(result.outputs.get('slug'), 'orientation');
  assert.ok(fs.existsSync(path.join(root, 'cohorts', '2026', 'events', 'orientation', 'index.md')));
  assert.ok(!fs.existsSync(path.join(root, 'cohorts', '2026', 'events', 'hijacked')));
});

test('new_event: a traversing year is refused and writes nothing', () => {
  for (const year of ['..', '2026/../..', '../../about']) {
    const root = fixtureTree();
    const before = fs.readFileSync(path.join(root, 'about', 'index.md'), 'utf8');
    const body =
      section('Cohort Year', year) + section('Event Title', 'Traversal') + section('Event ID', 'kickoff');

    const result = run('new_event_from_issue.mjs', { body, root });

    assert.equal(result.status, 1, `expected a refusal for ${year}`);
    assert.match(result.outputs.get('error') ?? '', /Refusing|required/);
    assert.equal(fs.readFileSync(path.join(root, 'about', 'index.md'), 'utf8'), before);
    assert.deepEqual(fs.readdirSync(path.join(root, 'cohorts', '2026', 'events')), ['kickoff']);
  }
});

test('new_event: a traversing event id stays inside the events folder', () => {
  const root = fixtureTree();
  const before = fs.readFileSync(path.join(root, 'about', 'index.md'), 'utf8');
  const body =
    section('Cohort Year', '2026') +
    section('Event Title', 'Traversal') +
    section('Event ID', '../../../about');

  const result = run('new_event_from_issue.mjs', { body, root });

  // The id is slugified before it is used, so it becomes a folder name here
  // rather than a path — and never escapes cohorts/2026/events/.
  assert.equal(fs.readFileSync(path.join(root, 'about', 'index.md'), 'utf8'), before);
  assert.deepEqual(fs.readdirSync(path.join(root, 'cohorts', '2026', 'events')).sort(), ['about', 'kickoff']);
  assert.equal(result.outputs.get('slug'), 'about');
});

test('new_event: control characters and quotes in a title survive the YAML round trip', () => {
  const root = fixtureTree();
  // A bell, a Unicode line separator and a lone carriage return: the old
  // hand-rolled escaper wrote all three raw, which either broke the document
  // or silently changed the value when it was read back.
  const title = 'Kickoff "quoted" \\ back\u0007 sep\u2028 cr\r end';
  const body =
    section('Cohort Year', '2026') + section('Event Title', title) + section('Event ID', 'controls');

  const result = run('new_event_from_issue.mjs', { body, root });

  assert.equal(result.status, 0, result.stderr);
  const data = frontMatterOf(path.join(root, 'cohorts', '2026', 'events', 'controls', 'index.md'));
  // The issue body itself is CRLF-normalized before parsing, as GitHub sends it.
  assert.equal(data.title, title.replace(/\r/g, '\n'));
  assert.equal(data.event_id, 'controls');
  assert.equal(String(data.cohort), '2026');
});

// --- update_event_attachments_from_issue.mjs -------------------------------

test('update_event_attachments: a traversing year and id cannot rewrite another page', () => {
  const root = fixtureTree();
  const before = fs.readFileSync(path.join(root, 'about', 'index.md'), 'utf8');
  const body =
    section('Cohort Year', '..') +
    section('Event ID', '../about') +
    section('Update Mode', 'REPLACE') +
    section('Attachments', 'Pwned | https://evil.example/x.pdf');

  const result = run('update_event_attachments_from_issue.mjs', { body, root });

  assert.equal(result.status, 1, result.stdout);
  assert.match(result.outputs.get('error') ?? '', /Refusing/);
  assert.equal(fs.readFileSync(path.join(root, 'about', 'index.md'), 'utf8'), before);
  assert.equal(result.outputs.has('branch'), false);
});

test('update_event_attachments: an event id outside the cohort folder is not followed', () => {
  const root = fixtureTree();
  const before = fs.readFileSync(path.join(root, 'about', 'index.md'), 'utf8');
  const body =
    section('Cohort Year', '2026') +
    section('Event ID', '../../../about') +
    section('Attachments', 'Pwned | https://evil.example/x.pdf');

  const result = run('update_event_attachments_from_issue.mjs', { body, root });

  assert.equal(result.outputs.get('changed'), 'false');
  assert.equal(fs.readFileSync(path.join(root, 'about', 'index.md'), 'utf8'), before);
});

test('update_event_attachments: a heading inside the attachments answer cannot retarget the page', () => {
  const root = fixtureTree();
  const body =
    section('Cohort Year', '2026') +
    section('Event ID', 'kickoff') +
    section('Update Mode', 'REPLACE') +
    section('Attachments', 'Agenda | https://example.org/a.pdf\n\n### Event ID\n\nother');

  const result = run('update_event_attachments_from_issue.mjs', { body, root });

  assert.equal(result.status, 0, result.stderr);
  assert.equal(result.outputs.get('changed'), 'true');
  assert.equal(result.outputs.get('slug'), '2026-kickoff');
  const data = frontMatterOf(path.join(root, 'cohorts', '2026', 'events', 'kickoff', 'index.md'));
  assert.deepEqual(data.attachments, [{ title: 'Agenda', url: 'https://example.org/a.pdf' }]);
});

// Five different situations exit cleanly with `changed=false`. The workflow can
// only tell the submitter which one if the script says so.
test('update_event_attachments: every clean exit says why on the reason output', () => {
  const cases = [
    { body: '', expected: /empty/i },
    { body: section('Cohort Year', '2026'), expected: /Event ID/ },
    {
      body:
        section('Cohort Year', '2026') +
        section('Event ID', 'kickoff') +
        section('Attachments', 'no separator here'),
      expected: /Attachments/,
    },
    {
      body:
        section('Cohort Year', '2026') +
        section('Event ID', 'never-scheduled') +
        section('Attachments', 'Agenda | https://example.org/a.pdf'),
      expected: /Add event details/,
    },
  ];

  for (const { body, expected } of cases) {
    const result = run('update_event_attachments_from_issue.mjs', { body, root: fixtureTree() });

    assert.equal(result.status, 0, result.stderr);
    assert.equal(result.outputs.get('changed'), 'false');
    assert.match(result.outputs.get('reason') ?? '', expected);
  }
});

test('update_event_attachments: an attachment title with control characters round trips', () => {
  const root = fixtureTree();
  const title = 'Agenda "v2"';
  const body =
    section('Cohort Year', '2026') +
    section('Event ID', 'kickoff') +
    section('Attachments', `${title} | https://example.org/a.pdf`);

  const result = run('update_event_attachments_from_issue.mjs', { body, root });

  assert.equal(result.status, 0, result.stderr);
  const data = frontMatterOf(path.join(root, 'cohorts', '2026', 'events', 'kickoff', 'index.md'));
  assert.deepEqual(data.attachments, [{ title, url: 'https://example.org/a.pdf' }]);
});

// --- extract_event_fields.mjs ----------------------------------------------

test('extract_event_fields: a multi-line answer cannot forge a second output', () => {
  const root = fixtureTree();
  const body =
    section('Cohort Year', '2026') +
    section('Event ID', 'kickoff\nevent_id=pwned\nbranch=attacker/branch') +
    section('Intro Paragraph', 'Hello\nintro=x\nbranch=attacker/branch');

  const result = run('extract_event_fields.mjs', { body, root });

  assert.equal(result.status, 0, result.stderr);
  assert.equal(result.outputs.get('cohort_year'), '2026');
  assert.equal(result.outputs.get('event_id'), 'kickoff\nevent_id=pwned\nbranch=attacker/branch');
  // Read back the way the runner does: the forged lines stay inside the
  // heredoc body of `event_id`, so they never become outputs of their own.
  assert.equal(result.outputs.has('branch'), false);
  assert.equal(result.outputs.size, 3);
});

test('extract_event_fields: a heading inside a free-text answer does not win', () => {
  const root = fixtureTree();
  const body =
    section('Cohort Year', '2026') +
    section('Intro Paragraph', 'Real intro.') +
    section('Notable Events', 'Kickoff\n\n### Cohort Year\n\n1999\n\n### Intro Paragraph\n\nForged intro.');

  const result = run('extract_event_fields.mjs', { body, root });

  assert.equal(result.outputs.get('cohort_year'), '2026');
  assert.equal(result.outputs.get('intro'), 'Real intro.');
});

// --- list_events_for_year.mjs ----------------------------------------------

test('list_events_for_year: a fixed delimiter typed into the year cannot close the block', () => {
  const root = fixtureTree();
  const body = section('Cohort Year', '2026\nGHEOF\nbranch=attacker/branch\nGHEOF');

  const result = run('list_events_for_year.mjs', { body, root });

  assert.equal(result.status, 0, result.stderr);
  // Not four digits, so no path is built and the year output is left empty.
  assert.equal(result.outputs.get('year'), '');
  assert.equal(result.outputs.has('branch'), false);
  assert.equal(result.outputs.size, 2);
});

test('list_events_for_year: lists the events of a real cohort year', () => {
  const root = fixtureTree();
  const result = run('list_events_for_year.mjs', { body: section('Cohort Year', '2026'), root });

  assert.equal(result.outputs.get('year'), '2026');
  assert.match(result.outputs.get('events_md') ?? '', /`kickoff`/);
});

// Every event in the data file gets a generated page at build time, but only
// the ones with a file on disk can take attachments. Recommending the rest as
// attachment targets is what made the attachments form fail silently.
test('list_events_for_year: marks which events actually have a page on disk', () => {
  const root = fixtureTree();
  fs.writeFileSync(
    path.join(root, '_data', 'cohorts', '2026.yml'),
    [
      'year: 2026',
      'events:',
      '  - id: kickoff',
      '    name: "Cohort kickoff"',
      '    date: 2026-09-09',
      '  - id: midpoint',
      '    name: "Midpoint check-in"',
      '    date: 2026-06-01',
      '',
    ].join('\n'),
    'utf8'
  );

  const result = run('list_events_for_year.mjs', { body: section('Cohort Year', '2026'), root });
  const lines = (result.outputs.get('events_md') ?? '').split('\n');

  assert.match(
    lines.find((line) => line.includes('`kickoff`')) ?? '',
    /has a details page/,
    'kickoff has an index.md in the fixture'
  );
  assert.match(
    lines.find((line) => line.includes('`midpoint`')) ?? '',
    /no details page yet/,
    'midpoint is only in the data file'
  );
});

File: test/scripts/event_scripts_test.rb
# frozen_string_literal: true

# Tests for the Ruby halves of the cohort automation:
# scripts/lib/issue_form.rb, scripts/update_schedule_from_issue.rb,
# scripts/preview_schedule_ids_from_issue.rb and scripts/scaffold_year.rb.
#
#   npm run test:ruby     (or: ruby -Itest test/scripts/event_scripts_test.rb)
#
# The two schedule scripts resolve their data file relative to their own
# directory, so the end-to-end cases copy the script (and the lib it needs) into
# a temporary tree — the real _data/cohorts/ is never touched.

require "minitest/autorun"
require "tmpdir"
require "fileutils"
require "yaml"

require_relative "../../scripts/lib/issue_form"

# Reads a $GITHUB_OUTPUT file the way the Actions runner does, so a test can
# tell a real output from a line that only looks like one inside a heredoc.
module OutputFile
  module_function

  def parse(raw)
    outputs = {}
    lines = raw.split("\n")
    index = 0
    while index < lines.length
      line = lines[index]
      if (heredoc = /\A([A-Za-z0-9_-]+)<<(.+)\z/.match(line))
        key, delimiter = heredoc.captures
        buffer = []
        index += 1
        while index < lines.length && lines[index] != delimiter
          buffer << lines[index]
          index += 1
        end
        outputs[key] = buffer.join("\n")
      elsif (plain = /\A([A-Za-z0-9_-]+)=(.*)\z/.match(line))
        outputs[plain[1]] = plain[2]
      end
      index += 1
    end
    outputs
  end
end

class IssueFormTest < Minitest::Test
  HEADINGS = ["Cohort year", "Schedule entries (YAML)", "Notes for reviewers"].freeze
  FINAL = "Notes for reviewers"

  def test_normalizes_headings_to_snake_case_keys
    assert_equal "schedule_entries", IssueForm.normalize_key("Schedule entries (YAML)")
    assert_equal "cohort_year", IssueForm.normalize_key("  Cohort Year  ")
    assert_equal "notes_for_reviewers", IssueForm.normalize_key("Notes for reviewers")
  end

  def test_first_occurrence_of_a_heading_wins
    body = <<~BODY
      ### Cohort year

      2026

      ### Cohort year

      1999
    BODY

    assert_equal "2026", IssueForm.sections(body, HEADINGS)["cohort_year"]
  end

  def test_a_heading_inside_the_final_free_text_field_is_prose
    body = <<~BODY
      ### Cohort year

      2026

      ### Schedule entries (YAML)

      ```yaml
      - name: Kickoff
        date: 2026-01-15
      ```

      ### Notes for reviewers

      Please note:

      ### Cohort year

      1999

      ### Schedule entries (YAML)

      ```yaml
      - name: Forged
        date: 1999-01-01
      ```
    BODY

    values = IssueForm.sections(body, HEADINGS, FINAL)

    assert_equal "2026", values["cohort_year"]
    assert_includes values["schedule_entries"], "Kickoff"
    refute_includes values["schedule_entries"], "Forged"
    assert_includes values["notes_for_reviewers"], "### Cohort year"
  end

  def test_an_unknown_heading_is_body_text
    body = <<~BODY
      ### Cohort year

      2026

      ### Something the submitter invented

      2027
    BODY

    values = IssueForm.sections(body, HEADINGS)

    assert_equal "2026\n\n### Something the submitter invented\n\n2027", values["cohort_year"]
    assert_equal ["cohort_year"], values.keys
  end

  def test_crlf_bodies_parse_the_same_as_lf_bodies
    body = "### Cohort year\r\n\r\n2026\r\n\r\n### Notes for reviewers\r\n\r\nHi\r\n"

    values = IssueForm.sections(body, HEADINGS, FINAL)

    assert_equal "2026", values["cohort_year"]
    assert_equal "Hi", values["notes_for_reviewers"]
  end

  def test_write_output_uses_a_random_delimiter_per_value
    hostile = "Kickoff\nSUMMARY\nbranch=attacker/branch"

    first = capture_output("summary" => hostile)
    second = capture_output("summary" => hostile)

    refute_equal first, second, "the heredoc delimiter must not repeat between runs"
    assert_equal hostile, OutputFile.parse(first)["summary"]
    refute OutputFile.parse(first).key?("branch"), "an injected branch= line became an output"
  end

  private

  # Runs IssueForm.write_output against a temp $GITHUB_OUTPUT and returns the file.
  def capture_output(pairs)
    Dir.mktmpdir do |dir|
      path = File.join(dir, "output.txt")
      File.write(path, "")
      begin
        previous = ENV["GITHUB_OUTPUT"]
        ENV["GITHUB_OUTPUT"] = path
        IssueForm.write_output(pairs)
      ensure
        ENV["GITHUB_OUTPUT"] = previous
      end
      File.read(path)
    end
  end
end

# End-to-end runs of the scripts themselves, in a throwaway tree.
class CohortScriptsTest < Minitest::Test
  ROOT = File.expand_path("../..", __dir__)

  def test_update_schedule_cannot_inject_a_branch_output
    hostile_name = "Kickoff\nSUMMARY\nbranch=attacker/branch\nchanged=true"
    body = <<~BODY
      ### Cohort year

      2999

      ### Schedule entries (YAML)

      ```yaml
      - name: #{hostile_name.dump}
        date: 2999-01-01
      ```
    BODY

    in_tree do |dir, output|
      status = run_script(dir, "update_schedule_from_issue.rb", body, output)
      outputs = OutputFile.parse(File.read(output))

      assert_equal 0, status
      assert_equal "true", outputs["changed"]
      assert_match %r{\Aschedule/2999-\d{14}\z}, outputs["branch"]
      assert_includes outputs["summary"], "branch=attacker/branch"
      assert_equal 1, YAML.safe_load(File.read(File.join(dir, "_data", "cohorts", "2999.yml")))["events"].length
    end
  end

  def test_update_schedule_ignores_a_heading_typed_into_the_notes
    body = <<~BODY
      ### Cohort year

      2999

      ### Schedule entries (YAML)

      ```yaml
      - name: Kickoff
        date: 2999-01-01
      ```

      ### Notes for reviewers

      Looks good.

      ### Schedule entries (YAML)

      ```yaml
      - name: Forged
        date: 1999-01-01
      ```
    BODY

    in_tree do |dir, output|
      status = run_script(dir, "update_schedule_from_issue.rb", body, output)
      events = YAML.safe_load(File.read(File.join(dir, "_data", "cohorts", "2999.yml")))["events"]

      assert_equal 0, status
      assert_equal ["Kickoff"], events.map { |event| event["name"] }
    end
  end

  def test_preview_matches_the_update_and_cannot_inject_an_output
    body = <<~BODY
      ### Cohort year

      2999

      ### Schedule entries (YAML)

      ```yaml
      - name: "Kickoff\\npreview_ids\\nbranch=attacker/branch"
        date: 2999-01-01
      ```

      ### Notes for reviewers

      ### Cohort year

      1999
    BODY

    in_tree do |dir, output|
      status = run_script(dir, "preview_schedule_ids_from_issue.rb", body, output)
      outputs = OutputFile.parse(File.read(output))

      assert_equal 0, status
      assert_equal "2999", outputs["year"]
      refute outputs.key?("branch"), "an injected branch= line became an output"
      assert_includes outputs["preview_ids"], "branch=attacker/branch"
    end
  end

  def test_scaffold_year_writes_front_matter_psych_can_parse
    hostile_intro = "Budget \#{1+1} plan     \"quoted\" \\ backslash"

    Dir.mktmpdir do |dir|
      FileUtils.mkdir_p(File.join(dir, "scripts", "lib"))
      FileUtils.cp(File.join(ROOT, "scripts", "scaffold_year.rb"), File.join(dir, "scripts"))
      # Whole lib, not a named list: the script requires from it and a new
      # shared file must not fail here as a missing require.
      FileUtils.cp(Dir[File.join(ROOT, "scripts", "lib", "*.rb")], File.join(dir, "scripts", "lib"))

      status = system(
        { "COHORT_YEAR" => "2999", "COHORT_INTRO" => hostile_intro },
        "ruby", File.join(dir, "scripts", "scaffold_year.rb"),
        out: File::NULL
      )

      assert status, "scaffold_year.rb failed"
      page = File.read(File.join(dir, "cohorts", "2999", "index.md"))
      front_matter = page.split(/^---$/)[1]
      data = YAML.safe_load(front_matter)

      assert_equal hostile_intro, data["intro"]
      assert_equal "Cohort 2999", data["title"]
      assert_equal 2999, data["year"]
    end
  end

  # Every refusal has to reach the issue, and the workflows can only quote what
  # reaches the `error` output — see IssueForm.fail.
  def test_update_schedule_reports_a_missing_cohort_on_the_error_output
    body = <<~BODY
      ### Cohort year

      2998

      ### Schedule entries (YAML)

      ```yaml
      - name: Kickoff
        date: 2998-01-01
      ```
    BODY

    in_tree do |dir, output|
      status = run_script(dir, "update_schedule_from_issue.rb", body, output)
      outputs = OutputFile.parse(File.read(output))

      assert_equal 1, status
      assert_includes outputs["error"], "2998"
      assert_includes outputs["error"], "Start a new cohort year"
      refute outputs.key?("changed")
    end
  end

  def test_update_schedule_reports_a_bad_date_on_the_error_output
    body = <<~BODY
      ### Cohort year

      2999

      ### Schedule entries (YAML)

      ```yaml
      - name: Kickoff
        date: next Tuesday
      ```
    BODY

    in_tree do |dir, output|
      status = run_script(dir, "update_schedule_from_issue.rb", body, output)

      assert_equal 1, status
      assert_includes OutputFile.parse(File.read(output))["error"], "2026-01-15"
    end
  end

  def test_scaffold_year_reports_a_bad_year_on_the_error_output
    in_scaffold_tree do |dir, output|
      status = run_scaffold(dir, "not-a-year", output)

      assert_equal 1, status
      assert_includes OutputFile.parse(File.read(output))["error"], "four-digit year"
    end
  end

  # The script never overwrites, so a second run writes nothing at all. The
  # workflow turns `created=false` into the "already exists" comment; without
  # the output there is a green run, no pull request and no reply.
  def test_scaffold_year_reports_that_an_existing_year_was_left_alone
    in_scaffold_tree do |dir, output|
      assert_equal 0, run_scaffold(dir, "2999", output)
      assert_equal "true", OutputFile.parse(File.read(output))["created"]

      second = File.join(dir, "second-output.txt")
      File.write(second, "")

      assert_equal 0, run_scaffold(dir, "2999", second)
      assert_equal "false", OutputFile.parse(File.read(second))["created"]
    end
  end

  private

  # Copies the schedule scripts into a temp tree with a cohort data file, so the
  # scripts write there instead of into the repository.
  def in_tree
    Dir.mktmpdir do |dir|
      FileUtils.mkdir_p(File.join(dir, "scripts", "lib"))
      FileUtils.mkdir_p(File.join(dir, "_data", "cohorts"))
      %w[update_schedule_from_issue.rb preview_schedule_ids_from_issue.rb].each do |script|
        FileUtils.cp(File.join(ROOT, "scripts", script), File.join(dir, "scripts"))
      end
      # Whole lib, not a named list: the scripts require from it and a new
      # shared file must not fail here as a missing require.
      FileUtils.cp(Dir[File.join(ROOT, "scripts", "lib", "*.rb")], File.join(dir, "scripts", "lib"))
      File.write(File.join(dir, "_data", "cohorts", "2999.yml"), "year: 2999\nevents: []\n")

      output = File.join(dir, "output.txt")
      File.write(output, "")
      yield dir, output
    end
  end

  def run_script(dir, script, body, output)
    ok = system(
      { "ISSUE_BODY" => body, "GITHUB_OUTPUT" => output },
      "ruby", File.join(dir, "scripts", script),
      out: File::NULL, err: File::NULL
    )
    ok ? 0 : 1
  end

  # scaffold_year.rb writes relative to its own directory, so it gets a tree of
  # its own — the real cohorts/ is never touched.
  def in_scaffold_tree
    Dir.mktmpdir do |dir|
      FileUtils.mkdir_p(File.join(dir, "scripts", "lib"))
      FileUtils.cp(File.join(ROOT, "scripts", "scaffold_year.rb"), File.join(dir, "scripts"))
      FileUtils.cp(Dir[File.join(ROOT, "scripts", "lib", "*.rb")], File.join(dir, "scripts", "lib"))

      output = File.join(dir, "output.txt")
      File.write(output, "")
      yield dir, output
    end
  end

  def run_scaffold(dir, year, output)
    ok = system(
      { "COHORT_YEAR" => year, "GITHUB_OUTPUT" => output },
      "ruby", File.join(dir, "scripts", "scaffold_year.rb"),
      out: File::NULL, err: File::NULL
    )
    ok ? 0 : 1
  end
end
